src/ng/q.js

```diff
@@ -223,22 +222,62 @@ function $QProvider() {
 
 function $$QProvider() {
   this.$get = ['$browser', '$exceptionHandler', function($browser, $exceptionHandler) {
-    return qFactory(function(callback) {
+    return qFactory(noop, function(callback, data) {
       $browser.defer(callback);
     }, $exceptionHandler);
   }];
 }
 
 /**
+ * @ngdoc service
+ * @name $qRaf
+ *
+ * @description
+ * `$qRaf` is a promise library that coordinates itself based on calls to `requestAnimationFrame`. The
+ * purpose of `$qRaf` is to allow animation-based code to be sequenced together without the need
+ * to hook into the digest cycle (which is how `$q` works).
+ *
+ * `$qRaf` is gauranteed to be delivered once one or more animation frames have passed. This means that
+ * even if the promise is resolved immediately then it will not be delivered until the next frame has passed.
+ * However if an animation does occur in between promise creation and resolution then the promise
+ * will be delivered immediately. This simple mechanism of flow control allows for animation callback
+ * code to avoid any unnecessary pauses. Any pause that occurs within an animation may lead to an
+ * unexpected reflow or page flicker and `$qRaf` aims to avoid this problem completely.
+ *
+ * Note that the API for `$qRaf` is the exact same as for `$q`.
+ *
+ * @param {function(function, function)} resolver Function which is responsible for resolving or
+ *   rejecting the newly created promise. The first parameter is a function which resolves the
+ *   promise, the second parameter is a function which rejects the promise.
+ *
+ * @returns {Promise} The newly created promise.
+ */
+function $QRafProvider() {
+  this.$get = ['$$rAF', '$exceptionHandler', function($$rAF, $exceptionHandler) {
+    return qFactory(
+      function(data) {
+        $$rAF(function() {
+          data.$$rafCleared = true;
+        });
+      },
+      function(callback, data) {
+        data.$$rafCleared ? callback() : $$rAF(callback);
```

Comment:
I believe that, here, `callback()` is only being called synchronously if an animation frame has already occurred between the creation of the deferred and the resolution of it.  Is that right? But I agree that this might not be enough asynchronicity.

The other versions of Q ($q and $$q) use `$evalAsync()` and `$browser.defer()` to ensure that the `callback` is not called until the calling function's stack has been unwound. In the case of `$$q`, it is always in a new browser loop; in the case of `$q` it may be a new browser loop or after the current digest completes.

@matsko - is the concern that waiting for the next animation frame is too late in some cases to prevent jank? My understanding of `requestAnimationFrame` is that you use it to ensure your callback occurs before the next repaint. If this is the case, then we should not need to call the `callback()` synchronously, right?

Removing this synchronous call would also allow us to ditch all the other API changes too, right?

@matsko - obviously you have added this for a reason but it would be good if you could clarify the benefits.